src/backend/distributed/planner/query_colocation_checker.c

```diff
@@ -361,8 +362,9 @@ CreateAllTargetListForRelation(Oid relationId, List *requiredAttributes)
 		}
 		else
 		{
+			int varAttNum = isMergeQuery ? attrNum : colAppendIdx++;
```

Comment:
I understand that there is an off-by-one error in the column values inserted into the table but I really would like understand the Postgres-internals fact that requires us doing something specific for MERGE while it's not needed for other types of commands.

Reply:
In Merge Command we should refer actual table attNum but in other cases we are building query like below which refers attnum of a subquery. not the actual table.
(SELECT t.a, NULL, NULL FROM (SELECT a FROM table) t).

That is the reason this is applicable only for merge command.